homeassistant/components/matrix/config_flow.py

```diff
@@ -0,0 +1,131 @@
+    """
+
+    try:
+        auth: MatrixAuthentication = MatrixAuthentication(
+            config_file=os.path.join(hass.config.path(), SESSION_FILE),
+            homeserver=data[CONF_HOMESERVER],
+            verify_ssl=data[CONF_VERIFY_SSL],
+            username=data[CONF_USERNAME],
+            password=data[CONF_PASSWORD],
+        )
+
+        client: MatrixClient = await hass.async_add_executor_job(auth._login)
+
+        # If no exception is thrown during logging in
+        token: str = ""
+        if hasattr(client, "token"):
+            # A new token will be assigned on first login
+            token = client.token
+        else:
+            # Use the token from the previous login if there is no token from the client
+            token = auth._auth_tokens[data[CONF_USERNAME]]
+
+        return {
+            "title": data[CONF_USERNAME],
+            CONF_ACCESS_TOKEN: token,
+        }
+
+    except (MatrixHttpLibError, MatrixRequestError):
+        # Re-raise so async_step_user can handle the exception
+        raise
+
+
+class ConfigFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow for Matrix."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: Optional[Dict[str, Any]] = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+
+        # Only allow single instance because
+        if self._async_current_entries():
+            return self.async_abort(reason="already_configured")
+
+        errors: Dict[str, str] = {}
+
+        if user_input is not None:
+            try:
+                info: Dict[str, str] = await validate_input(self.hass, user_input)
+
+            except MatrixHttpLibError:
+                # Network error
+                errors[CONF_BASE] = "cannot_connect"
+
+            except MatrixRequestError as ex:
+                # Error code definitions: https://spec.matrix.org/latest/client-server-api/#standard-error-response
+                error_code: Final[str] = json.loads(ex.content)["errcode"]
+
+                if error_code in ("M_FORBIDDEN", "M_UNAUTHORIZED"):
+                    errors[CONF_PASSWORD] = "invalid_auth"
+                elif error_code in ("M_INVALID_USERNAME", "M_USER_DEACTIVATED"):
+                    errors[CONF_USERNAME] = "invalid_auth"
+                elif error_code in ("M_UNKNOWN_TOKEN", "M_MISSING_TOKEN"):
+                    errors[CONF_BASE] = "invalid_access_token"
+                else:
+                    errors[CONF_BASE] = "unknown"
+
+            else:
+                # Use access token as unique id as it can't be duplicated
+                await self.async_set_unique_id(info[CONF_ACCESS_TOKEN])
+                self._abort_if_unique_id_configured()
+
+                # CONF_ROOMS: set empty because we don't have any
+                user_input[CONF_ROOMS] = []
+
+                # CONF_COMMANDS: set empty because we don't have any
+                user_input[CONF_COMMANDS] = []
+
```

Comment:
Why do we set these in the entry data?

Reply:
Will address after CONFIG_SCHEMA is marked deprecated.